designs/EmbeddedReadmesTechSpec.md

```diff
@@ -0,0 +1,102 @@
+
+```
+AddElementIfNotNull(elem, ns, "readme", metadata.Readme);
+```
+
+**If no Readme property is present in the source file, but a readme.md file exists in the base directory**
+
+Once the properties and files have been processed, if the PackageBuilder object's Readme property is null, we need to look through the base directory for a "readme.md" file. If such a file exists, we must include it in the package and and populate the Readme property with "readme.md".
+
+```
+public void CheckForReadme(string basePath)
+{
+    string readmeName = "readme.md";
+    List<PhysicalPackageFile> searchFiles = ResolveSearchPattern(basePath, readmeName, string.Empty, false).ToList();
+
+    if (searchFiles.Any())
+    {
+        Readme = readmeName;
+
+        if (!Files.Contains(searchFiles[0]))
+        {
+            Files.Add(searchFiles[0]);
+        }
+    }
+}
+```
+
+This CheckForReadme method in [PackageBuilder.cs](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs) is called [here](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs#L725) and [here](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs#L726), based on whether the package is being built using a *nuspec* or *csproj* file.
+
+**Validation and Errors**
+
+To validate the Readme property and the corresponding file when `pack` is called, the ValidateReadmeFile method in [PackageBuilder.cs](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs) checks that:
+* The file extension matches a Markdown file
+* The file specified in the Readme property exists in the package's files
+* The file can be opened
+* The file size does not exceed 1 MB
+* The file is not empty
+
+Each of these errors has a corresponding log code (NU5038 - NU5042) and error message defined in [NuGetLogCode.cs](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs) and [NuGetResources.resx](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Resources/NuGetResources.resx) respectively, and the appropriate support must be added in [NuGetResources.Designer.cs](https://github.com/NuGet/NuGet.Client/blob/0f8ad8263539cb9bc69c441569453c1da98fb4cc/src/NuGet.Core/NuGet.Packaging/PackageCreation/Resources/NuGetResources.Designer.cs).
+
+## Future Work
+
+* Adding readme support through the VS Project Properties UI
+* Adding a `View Documentation` link to the package readme in the PM UI
+
+## Open Questions
+
+* Should we warn users if the Readme property is empty?
```

Comment:
If they don't have a readme at all? 
I'd vote no for now. 
Lots of customer elevate warnings to errors and while a good quality package will have a readme, I'd rather that get added during the actual high quality package effort (@chgill-MSFT, @JonDouglas will know more).